Add platform setup and teardown operations in the examples

k-stachowiak opened the pull request:
This PR adds the platform setup and teardown operations to the example programs. In some cases the flow of the programs was modified to ensure the teardown procedure call regardless of the reason of the program termination.

## Status
**READY**

## Requires Backporting
This is not a bugfix, however if the platforms that require explicit setup are supposed to support older Mbed TLS versions, backports should be performed to the relevant branches.

## Migrations
NO

## Additional comments
This changes should be performed in all the programs that are expected to work with the platforms that require explicit setup.

## Todos
- [ ] Tests
- [ ] Documentation
- [ ] Changelog updated
- [ ] Backported

## Preceding PRs:
- [x] #2595
- [x] #2601
- [x] #2604

sbutcher-arm commented on programs/test/udp_proxy.c:
```diff
@@ -613,11 +616,16 @@ int main( int argc, char *argv[] )
 
+    mbedtls_platform_context platform_ctx;
+
+    mbedtls_platform_setup( &platform_ctx );
```
I think we're making the code unnecessarily more complex by placing part of the initialisation before the call to `get_options()`. Because of that, `get_options()` and `exit_usage()` now have to provide the same exit path to ensure `mbedtls_platform_teardown()` is called.

In my view, we should analyse and parse the options *before* we begin platform initialisation. What purpose is there to initialising the platform and network before we know whether we can execute?

I think `get_options()` should be called before `mbedtls_platform_setup()` and `mbedtls_net_init()`. If you do that you also didn't have to rewrite the options parsing and usage, which I'd revert.

k-stachowiak replied:
In principle I assumed that the platform initialization may potentially impact any standard function, therefore if the users expect the platform specific code to be always enabled for the given platform the initialization function should be called as soon as possible.

For example if the platform initialization impacts the file system access and the options parsing involves opening files, it will depend on the platform initialization having been called beforehand.

Please let me know if I'm mistaken. I will start reverting relevant changes immediately (this has been done for more than one example), but maybe it is a valid approach after all?

RonEld commented on programs/pkey/pk_sign.c:
```diff
@@ -70,6 +73,13 @@ int main( int argc, char *argv[] )
+    if( ( ret = mbedtls_platform_setup( &platform_ctx ) ) != 0 )
+    {
+        mbedtls_printf( " failed\n  ! mbedtls_platofrm_setup returned %d\n", ret );
+        goto exit;
```
This should be changed to `return ret`. If you go to exit here, you will free uninitialized resources, which might lead to undefined behavior, depending on implementation.

RonEld commented on programs/pkey/pk_verify.c:
```diff
@@ -57,11 +57,22 @@ int main( int argc, char *argv[] )
+    if( ( ret = mbedtls_platform_setup( &platform_ctx ) ) != 0 )
+    {
+        mbedtls_printf( " failed\n  ! mbedtls_platform_setup returned %d\n", ret );
+        goto exit;
```
This should be changed to `return ret`. If you go to exit here, you will free uninitialized resources, which might lead to undefined behavior, depending on implementation.

RonEld commented on programs/pkey/rsa_decrypt.c:
```diff
@@ -72,6 +75,16 @@ int main( int argc, char *argv[] )
+        mbedtls_printf( " failed\n  ! mbedtls_platofrm_setup returned %d\n",
+                        return_val );
+        exit_val = MBEDTLS_EXIT_FAILURE;
+        goto exit;
```
This should be changed to `return ret`. If you go to exit here, you will free uninitialized resources, which might lead to undefined behavior, depending on implementation.

RonEld commented on programs/pkey/rsa_encrypt.c:
```diff
@@ -71,6 +74,16 @@ int main( int argc, char *argv[] )
+        mbedtls_printf( " failed\n  ! mbedtls_platform_setup returned %d\n",
+                        return_val );
+        exit_val = MBEDTLS_EXIT_FAILURE;
+        goto exit;
```
This should be changed to `return ret`. If you go to exit here, you will free uninitialized resources, which might lead to undefined behavior, depending on implementation.

RonEld commented on programs/pkey/rsa_genkey.c:
```diff
@@ -70,6 +73,14 @@ int main( void )
+    if( ( ret = mbedtls_platform_setup( &platform_ctx ) ) != 0 )
+    {
+        mbedtls_printf( " failed\n  ! mbedtls_platform_setup returned %d\n", ret );
+        goto exit;
```
This should be changed to `return ret`. If you go to exit here, you will free uninitialized resources, which might lead to undefined behavior, depending on implementation.

RonEld commented on programs/pkey/rsa_sign.c:
```diff
@@ -57,12 +57,23 @@ int main( int argc, char *argv[] )
+    if( ( ret = mbedtls_platform_setup( &platform_ctx ) ) != 0 )
+    {
+        mbedtls_printf( " failed\n  ! mbedtls_platform_setup returned %d\n\n", ret );
+        goto exit;
```
This should be changed to `return ret`. If you go to exit here, you will free uninitialized resources, which might lead to undefined behavior, depending on implementation.

RonEld commented on programs/random/gen_entropy.c:
```diff
@@ -50,18 +50,35 @@ int main( int argc, char *argv[] )
+    if( mbedtls_platform_setup( &platform_ctx ) != 0 )
+    {
+        mbedtls_printf( "failed!\n" );
+        return( 1 );
```
finally!:)

RonEld commented on programs/random/gen_random_ctr_drbg.c:
```diff
@@ -53,21 +53,37 @@ int main( int argc, char *argv[] )
         mbedtls_fprintf( stderr, "usage: %s <output filename>\n", argv[0] );
+#if defined(MBEDTLS_PLATFORM_C)
+        mbedtls_platform_teardown( &platform_ctx );
+#endif
```
not part of the scope of this PR, but I would also call `mbedtls_ctr_drbg_free`

RonEld commented on programs/random/gen_random_ctr_drbg.c:
```diff
@@ -53,21 +53,37 @@ int main( int argc, char *argv[] )
         mbedtls_printf( "failed to open '%s' for writing.\n", argv[1] );
+#if defined(MBEDTLS_PLATFORM_C)
+        mbedtls_platform_teardown( &platform_ctx );
+#endif
```
not part of the scope of this PR, but I would also call `mbedtls_ctr_drbg_free`

RonEld commented on programs/ssl/ssl_fork_server.c:
```diff
@@ -100,13 +100,24 @@ int main( void )
+        mbedtls_printf( " failed!  mbedtls_platform_setup returned %d\n\n", ret );
+        goto exit;
+    }
+#endif
```
for `ssl_fork_server`, I would expect this to be called once, for platform initialization, like you did, but I would also expect every child process to have access to the platform context. I would like to protect the context with a semaphore, and have the context sent as a shared resource.

k-stachowiak replied:
As discussed off line I raised an issue regarding this here: https://github.com/ARMmbed/mbedtls/issues/1583.

RonEld commented on programs/ssl/ssl_pthread_server.c:
```diff
@@ -331,6 +334,14 @@ int main( void )
+        mbedtls_printf( " failed\n  !  mbedtls_platform_setup returned %d\n\n", ret );
+        goto exit;
+    }
+#endif
```
for `ssl_pthread_server`, I would add the platform context to `thread_info_t` , and protect it with a mutex

RonEld commented on programs/util/pem2der.c:
```diff
@@ -185,6 +185,14 @@ int main( int argc, char *argv[] )
+    {
+        mbedtls_printf( " failed\n  !  mbedtls_platform_setup returned %d\n\n", ret );
+        goto exit;
+    }
```
do we need platform initialization for this example? We don't have an alternative implementation for PEM \ DER parsing, and I don't see it happenning in the near future

k-stachowiak replied:
I am not sure about this approach. I can imagine someone wanting to run any on a specific platform. In such case all the examples should be prepared for that. Do you think otherwise?

RonEld replied:
Even if this example runs on the platform, no cryptographic  platform initialization needs to be done for PEM \ DER, as it is SW only code.

RonEld commented on programs/test/udp_proxy.c:
```diff
@@ -613,11 +616,21 @@ int main( int argc, char *argv[] )
+        goto exit;
+    }
+#endif
+
```
Do we need platform initialization here? This is only a udp proxy, without any reason to have aplatform context here ( no cryptography, and only networking API)

RonEld commented on programs/util/strerror.c:
```diff
@@ -55,11 +55,20 @@ int main( int argc, char *argv[] )
+        mbedtls_printf( "  Failed initializing platform.\n" );
+        goto exit;
+    }
+#endif
```
This is a utility program, that translates errors to string. No need for platform context initialization here.

k-stachowiak commented:
>I have left a comment, which is relevant for most applications:
I would consider moving this statement before the windows getchar part, as I would expect the "press any key to continue" to be the last part, and if, for example, I am with a debugger, I would like to see all resources free at this time.

This seems like a reasonable behavior, especially that in `WIN32` context one can be pretty sure that `getchar()` works even after the call to `platform_teardown{}`. However this is not the way the programs worked so far; at least not all of them. Please note that the `ecdh_curve25519` program has the `getchar()` called before the cryptographic resources being released. Please confirm that this change should be made within this PR. I am afraid this change may not get approved by other reviewers.

RonEld commented:
@k-stachowiak The idea for the `mbedtls_platform_context` is to hold the platform context for Mbed TLS, meaning, the cryptography part. so I don't think having the platform context setup after the options affect other platform initialization. For example, other memory management ( e.g. I/O , memory management ) is part of the platform driver code, and not part of the Mbed TLS plafrom code.

RonEld commented:
@k-stachowiak I think we should have consistency in our code.  In your example (`ecdh_curve25519 `), although it's a bit out of scope, will it be possible to move the calls for cryptography resource freeing before the `getchar` call?

mazimkhan commented on programs/aes/aescrypt2.c:
```diff
@@ -87,6 +87,9 @@ int main( int argc, char *argv[] )
     unsigned char diff;
 
+#if defined(MBEDTLS_PLATFORM_C)
+    mbedtls_platform_context platform_ctx;
```
In case of _ALT implementations where this context will be used by the platform specific code, it may be useful to create it global instead of a stack variable. We don't know how big it could be and if it will fit the embedded application stack. Secondly, platform specific code may want to access it as an ```extern``` object.

k-stachowiak replied:
@mazimkhan Do you advise doing this for all the examples?
@RonEld What do you think about such use case? It sound's reasonable, but I'd like to consult this with more reviewers?

mazimkhan replied:
Yes as the argument is general.

k-stachowiak replied:
It seems to me that there is no immediate need to make platform context global in this PR. @mazimkhan do you agree?

mazimkhan replied:
Ok if there is an issue #1200 addressing scope, memory requirement and ownership of this object then I am fine leaving it as is.

k-stachowiak commented on programs/ssl/ssl_pthread_server.c:
```diff
@@ -105,6 +105,10 @@ static void my_mutexed_debug( void *ctx, int level,
 typedef struct {
+#if defined(MBEDTLS_PLATFORM_C)
+    mbedtls_platform_context platform_ctx;
+    mbedtls_threading_mutex_t platform_ctx_mutex;
```
I can't find any other mutex in this program except for the `debug_mutex`, which I don't think is a proper resource for locking the platform context access. Please let me know if I am missing something.

mazimkhan replied:
I don't see use of mutex here. ```platform_ctx``` is added to the ```thread_info_t``` struct that is instantiated per thread. Hence, this mutex is not adding any value here itself. Secondly, if an _ALT implementation implements ```mbedtls_platform_context ``` in thread unsafe way then it should that implementation's responsibility to ensure thread safety with mutex(s).

k-stachowiak commented:
> I don't see use of mutex here. platform_ctx is added to the thread_info_t struct that is instantiated per thread. Hence, this mutex is not adding any value here itself. Secondly, if an _ALT implementation implements mbedtls_platform_context in thread unsafe way then it should that implementation's responsibility to ensure thread safety with mutex(s).

@mazimkhan The `thread_info_t` structure is actually shared between the threads and it is the `pthread_info_t` that is thread-local. However I still agree that this should still be reverted for the same reason for which we don't store the platform context as a global variable.

k-stachowiak commented:
@mazimkhan I removed the platform context from the shared structure as I think it is better like this before a common strategy for sharing it as been decided. @RonEld Do you agree with this idea?

mazimkhan commented on programs/test/selftest.c:
```diff
@@ -387,11 +397,15 @@ int main( int argc, char *argv[] )
+#endif
 #if defined(_WIN32)
         mbedtls_printf( "  Press Enter to exit this program.\n" );
         fflush( stdout ); getchar();
```
This has been moved out of above ```if``` block. Please fix indentation.

k-stachowiak commented on programs/ssl/ssl_fork_server.c:
```diff
@@ -402,12 +417,14 @@ int main( void )
-
+#if defined(MBEDTLS_PLATFORM_C)
+    if( is_parent )
+        mbedtls_platform_teardown( &platform_ctx );
```
The idea here is to only call the platform teardown operation once, as we only call the setup once. As far as I understand this involves setting up hardware resources, and hence they should only be deinitialized/released once afterwards. I can see a problem when the platform context has some software resources associated, that would have to be deinitialized for each process.

I think that the risk of multiple teardown calls from the HW perspective is low, as if this ever poses a threat, the teardown procedure implementation may contain necessary checks. OTOH it seems possible that runtime resources will be stored in the context object, therefore I agree that this should be changed.

RonEld commented:
@k-stachowiak I partially agree.
The platform context should be accessed by the threads, this means the context should be sent to the thread through `thread_info_t`(probably as a pointer), unless you see a different way to do so.
As for the mutex part, since this application is generating the threads, and this application is the *owner* of the resource (doesn't matter what internal implementation this resource has), this application should be responsible for protecting its resource. An alternative implementer can add a mutex of their own, for their reasons, but the internal implementation of the context should interest the application. Although a mutex is not used at the moment, I would still add a mutex in comment, mentioning it should be used once #1200 is resolved.

k-stachowiak commented:
@RonEld I can see two strong scenarios here:

1. If the thread actually accessed the platform context, then it would have to be shared and the mutex would be necessary.

2. Since the threads don't need the access to the platform context it doesn't have to be shared. Also since it is not shared the unused mutex is not useful as well. Having it shared would make sense as an example, but since it has not been decided yet, how the sharing is supposed to be done it may make no sense to implement sharing now.

Doing something between the above two seems a bit vague to me. My main problem is declaring a mutex, initializing and deinitializing it, but not using it even once. Also we make extra effort to put the context object in the shared structure while the context is shared nowhere. I'm afraid this may confuse the users more than it helps them.

mazimkhan commented:
@k-stachowiak @RonEld My understanding is as follows:
- We create ```platform_ctx``` in ```main``` and register it with the library using ```mbedtls_platform_setup```. This way it gets penetrated in the library for use.
- Any mutex that we create to make ```platform_ctx``` thread safe should be accessible at the places that use ```platform_ctx```. Hence creating the mutex here does not help. It should either supplied via ```mbedtls_platform_setup```. Or can be part of ```mbedtls_platform_ctx``` itself, making it implementation specific. 
- We can wait for #1200 that will add more clarity to it.
- Given the current API ```mbedtls_platform_setup``` I don't find it useful to have a mutex here.

RonEld commented:
@mazimkhan the context is not being registered in `mbedtls_platform_setup`. If it was being registered, it meant that there was some sort of global variable holding it. If there was some sort of global variable holding it, which complexes the implementation,  there was no point in having the context as a parameter to `mbedtls_platform_setup`, since it's already global.
Of course, any alternative implementer can do what they wish, but it's not part of the designed API.
I agree that creating a mutex at the moment does not make sense, this is why I asked to add a comment about the necessity, once #1200 is addressed.
As I mentioned previously, since the Platform context is a resource owned by this application, it should be protected by this application. The internals of the resource which are platform specific, can be oreitected in a platform specific way, although I prefer not having too much mutexes.

@k-stachowiak as for your two use cases, as I see it, number 1 is the only use case once #1200 is addressed. Hence the need of the mutex. number two is unfortunately the current situation, where the platform context is not really needed anywhere, except in the `mbedtls_platform_setup` API, which makes it redundant.

mazimkhan commented:
@RonEld thanks for clarifying that it is not registered and it is only owned by the application. I will leave it to further discussion in #1200. I don't mind putting a comment about future thread safety consideration.

k-stachowiak commented:
@RonEld , @mazimkhan Could you summarize your conclusions in terms of the review? Do you find the submitted solution acceptable? My last interpretation of the discussion is that there was no consensus, which is needed to proceed with the PR.

I favor leaving the implementation as is, or adding a comment next to the `mbedtls_platform_context` declaration, that if it is to be shared between the threads it should be registered in the thread info structure and guarded by a mutex.

mazimkhan commented:
@k-stachowiak thanks for working all requested changes.

sbutcher-arm commented on ChangeLog:
```diff
@@ -93,6 +93,9 @@ Changes
    * Improve robustness of mbedtls_ssl_derive_keys against the use of
      HMAC functions with non-HMAC ciphersuites. Independently contributed
      by Jiayuan Chen in #1377. Fixes #1437.
+   * Add calls to platform setup and teardown functions to all the example
```
The ChangeLog entry reads a little awkwardly.
> Add calls to platform setup and teardown functions to all the example programs for which this is relevant.

Can I suggest:

> Add calls to the platform abstraction setup and teardown functions, into the example programs.

sbutcher-arm commented on programs/pkey/gen_key.c:
```diff
@@ -192,13 +192,24 @@ int main( int argc, char *argv[] )
+#if defined(MBEDTLS_PLATFORM_C)
+    if( ( ret = mbedtls_platform_setup( &platform_ctx ) ) != 0 )
+    {
+        mbedtls_printf( " failed\n  ! mbedtls_platform_setup returned %d\n", ret );
```
Shouldn't we display `-ret` for consistency with line 108?

sbutcher-arm commented on programs/pkey/key_app.c:
```diff
@@ -88,9 +88,20 @@ int main( int argc, char *argv[] )
+#if defined(MBEDTLS_PLATFORM_C)
+    if( ( ret = mbedtls_platform_setup( &platform_ctx ) ) != 0 )
+    {
+        mbedtls_printf( " failed\n  !  mbedtls_platform_setup returned %d\n", ret );
```
Shouldn't we display `-ret` for consistency with line 184?

sbutcher-arm commented on programs/pkey/key_app_writer.c:
```diff
@@ -194,9 +194,20 @@ int main( int argc, char *argv[] )
+#if defined(MBEDTLS_PLATFORM_C)
+    if( mbedtls_platform_setup( &platform_ctx ) != 0 )
+    {
+        mbedtls_printf( " failed\n  ! could not initialize platform\n\n" );
```
For consistency, why don't we output `ret` here?

sbutcher-arm commented on programs/pkey/pk_decrypt.c:
```diff
@@ -69,6 +72,13 @@ int main( int argc, char *argv[] )
+#if defined(MBEDTLS_PLATFORM_C)
+    if( ( ret = mbedtls_platform_setup( &platform_ctx ) ) != 0 )
+    {
+        mbedtls_printf( " failed\n  ! mbedtls_platform_setup returned %d\n", ret );
```
Shouldn't we display `-ret` for consistency with line 138?

sbutcher-arm commented on programs/pkey/pk_encrypt.c:
```diff
@@ -61,13 +61,24 @@ int main( int argc, char *argv[] )
+#if defined(MBEDTLS_PLATFORM_C)
+    if( ( ret = mbedtls_platform_setup( &platform_ctx ) ) != 0 )
+    {
+        mbedtls_printf( " failed\n  ! mbedtls_platform_setup returned %d\n", ret );
```
Shouldn't we display `-ret` for consistency with line 126?

sbutcher-arm commented on programs/pkey/pk_sign.c:
```diff
@@ -70,6 +73,13 @@ int main( int argc, char *argv[] )
+#if defined(MBEDTLS_PLATFORM_C)
+    if( ( ret = mbedtls_platform_setup( &platform_ctx ) ) != 0 )
+    {
+        mbedtls_printf( " failed\n  ! mbedtls_platofrm_setup returned %d\n", ret );
```
Shouldn't we display `-ret` for consistency with line 126?

sbutcher-arm commented on programs/pkey/pk_verify.c:
```diff
@@ -57,11 +57,22 @@ int main( int argc, char *argv[] )
+#if defined(MBEDTLS_PLATFORM_C)
+    if( ( ret = mbedtls_platform_setup( &platform_ctx ) ) != 0 )
+    {
+        mbedtls_printf( " failed\n  ! mbedtls_platform_setup returned %d\n", ret );
```
Shouldn't we display `-ret` for consistency with line 122?

sbutcher-arm commented on programs/pkey/rsa_sign.c:
```diff
@@ -57,12 +57,23 @@ int main( int argc, char *argv[] )
+#if defined(MBEDTLS_PLATFORM_C)
+    if( ( ret = mbedtls_platform_setup( &platform_ctx ) ) != 0 )
+    {
+        mbedtls_printf( " failed\n  ! mbedtls_platform_setup returned %d\n\n", ret );
```
Shouldn't we display `-ret` for consistency with line 151?

k-stachowiak commented:
@sbutcher-arm, @gilles-peskine-arm: If @RonEld or another second reviewer has approved the changes this will be ready for merge. The backport PR was made by cherry picking the commits from this one, so the changes are basically the same there.

sbutcher-arm commented on programs/pkey/rsa_verify.c:
```diff
@@ -56,11 +56,21 @@ int main( int argc, char *argv[] )
+#if defined(MBEDTLS_PLATFORM_C)
+    if( ( ret = mbedtls_platform_setup( &platform_ctx ) ) != 0 )
+    {
+        mbedtls_printf( " failed\n  ! mbedtls_platform_setup returned %d\n\n", ret );
```
Shouldn't we display `-ret` for consistency with other errors?

sbutcher-arm commented on programs/ssl/dtls_client.c:
```diff
@@ -99,6 +102,14 @@ int main( int argc, char *argv[] )
+#if defined(MBEDTLS_PLATFORM_C)
+    if( ( ret = mbedtls_platform_setup( &platform_ctx ) ) != 0 )
+    {
+        mbedtls_printf( " failed\n  ! mbedtls_platform_setup returned %d\n\n", ret );
```
Shouldn't we display `-ret` for consistency with other errors?

sbutcher-arm commented on programs/ssl/ssl_pthread_server.c:
```diff
@@ -331,6 +334,14 @@ int main( void )
+#if defined(MBEDTLS_PLATFORM_C)
+    if( ( ret = mbedtls_platform_setup( &platform_ctx ) ) != 0 )
+    {
+        mbedtls_printf( " failed\n  !  mbedtls_platform_setup returned %d\n\n", ret );
```
Shouldn't we display `-ret` for consistency with other errors?

sbutcher-arm commented on programs/ssl/ssl_server.c:
```diff
@@ -107,6 +110,14 @@ int main( void )
+#if defined(MBEDTLS_PLATFORM_C)
+    if( ( ret = mbedtls_platform_setup( &platform_ctx ) ) != 0 )
+    {
+        mbedtls_printf( " failed\n  !  mbedtls_platform_setup returned %d\n\n", ret );
```
Shouldn't we display `-ret` for consistency with other errors?

sbutcher-arm commented on programs/x509/cert_app.c:
```diff
@@ -160,6 +163,14 @@ int main( int argc, char *argv[] )
+#if defined(MBEDTLS_PLATFORM_C)
+    if( ( ret = mbedtls_platform_setup( &platform_ctx ) ) != 0 )
+    {
+        mbedtls_printf( " failed\n  !  mbedtls_platform_setup returned %d\n\n", ret );
```
Shouldn't we display `-ret` for consistency with other errors?

sbutcher-arm commented on programs/ssl/ssl_client2.c:
```diff
@@ -534,6 +537,14 @@ int main( int argc, char *argv[] )
+#if defined(MBEDTLS_PLATFORM_C)
+    if( mbedtls_platform_setup( &platform_ctx ) != 0 )
+    {
+        mbedtls_printf( "Failed initializing platform.\n" );
```
Error message is inconsistent with other error messages in the app. e.g. `" failed\n  ! mbedtls_ssl_handshake returned -0x%x\n\n"` at line 1972.

sbutcher-arm commented on programs/util/pem2der.c:
```diff
@@ -275,12 +275,10 @@ int main( int argc, char *argv[] )
 
 exit:
     free( pem_buffer );
-
```
Why are we removing lines here? The file is otherwise untouched.

AndrzejKurek commented:
Added all of the changes requested by Simon.
@sbutcher-arm, @RonEld, please re-review.

AndrzejKurek commented:
Added changes as requested in the backport PR. @RonEld - please re-review.

AndrzejKurek commented:
The only failing test is the timing test suite:
10:08:44 [gmake-clang-freebsd] test_suite_timing ................................................. FAIL

sbutcher-arm commented on programs/pkey/rsa_verify.c:
```diff
@@ -56,11 +56,21 @@ int main( int argc, char *argv[] )
+#if defined(MBEDTLS_PLATFORM_C)
+    if( ( ret = mbedtls_platform_setup( &platform_ctx ) ) != 0 )
+    {
+        mbedtls_printf( " failed\n  ! mbedtls_platform_setup returned %d\n\n", -ret );
```
The return value is displayed as a decimal here, but as hexadecimal at line 143. We should be self-consistent within the same program.

sbutcher-arm commented on programs/pkey/rsa_verify_pss.c:
```diff
@@ -61,11 +61,22 @@ int main( int argc, char *argv[] )
+#if defined(MBEDTLS_PLATFORM_C)
+    if( ( ret = mbedtls_platform_setup( &platform_ctx ) ) != 0 )
+    {
+        mbedtls_printf( " failed\n  !  mbedtls_platform_setup returned %d\n\n", -ret );
```
The return value is displayed as a negative integer here, but as positive at line 136. We should be self-consistent within the same program.

sbutcher-arm commented on programs/random/gen_random_ctr_drbg.c:
```diff
@@ -53,21 +53,39 @@ int main( int argc, char *argv[] )
+#if defined(MBEDTLS_PLATFORM_C)
+    if( ( ret = mbedtls_platform_setup( &platform_ctx ) ) != 0 )
+    {
+        mbedtls_printf( " failed\n  !  mbedtls_platform_setup returned %d\n\n", -ret );
```
The return value is displayed as a negative here, but as positive at line 92. We should be self-consistent within the same program.

sbutcher-arm commented on programs/pkey/rsa_genkey.c:
```diff
@@ -70,6 +73,14 @@ int main( void )
+#if defined(MBEDTLS_PLATFORM_C)
+    if( ( ret = mbedtls_platform_setup( &platform_ctx ) ) != 0 )
+    {
+        mbedtls_printf( " failed\n  !  mbedtls_platform_setup returned %d\n\n", -ret );
```
We're displaying a negative return value here, yet at line 122 and onwards, are displaying it without the negative. We should be self-consistent within the same program.

sbutcher-arm commented on programs/pkey/rsa_sign.c:
```diff
@@ -57,12 +57,23 @@ int main( int argc, char *argv[] )
+#if defined(MBEDTLS_PLATFORM_C)
+    if( ( ret = mbedtls_platform_setup( &platform_ctx ) ) != 0 )
+    {
+        mbedtls_printf( " failed\n  ! mbedtls_platform_setup returned %d\n\n", -ret );
```
We're displaying a negative return value here, yet at line 105 and onwards, are displaying it without the negative. We should be self-consistent within the same program.

sbutcher-arm commented on programs/pkey/rsa_sign_pss.c:
```diff
@@ -70,6 +73,13 @@ int main( int argc, char *argv[] )
+#if defined(MBEDTLS_PLATFORM_C)
+    if( ( ret = mbedtls_platform_setup( &platform_ctx ) ) != 0 )
+    {
+        mbedtls_printf( " failed\n  !  mbedtls_platform_setup returned %d\n\n", -ret );
```
We're displaying a negative return value here, yet at line 137 and onwards, are displaying it without the negative. We should be self-consistent within the same program.

sbutcher-arm commented on programs/ssl/dtls_client.c:
```diff
@@ -99,6 +102,14 @@ int main( int argc, char *argv[] )
+#if defined(MBEDTLS_PLATFORM_C)
+    if( ( ret = mbedtls_platform_setup( &platform_ctx ) ) != 0 )
+    {
+        mbedtls_printf( " failed\n  ! mbedtls_platform_setup returned %d\n\n", -ret );
```
We're displaying a decimal return value here, yet at line 293, are displaying it as a hexadecimal. We should be self-consistent within the same program.

sbutcher-arm commented on programs/ssl/dtls_server.c:
```diff
@@ -110,6 +113,13 @@ int main( void )
+#if defined(MBEDTLS_PLATFORM_C)
+    if( ( ret = mbedtls_platform_setup( &platform_ctx ) ) != 0 )
+    {
+        printf( " failed\n  !  mbedtls_platform_setup returned %d\n\n", -ret );
```
We're displaying a negative return value here, yet at line 361, are displaying it without the negative. We should be self-consistent within the same program.

sbutcher-arm commented on programs/ssl/ssl_client1.c:
```diff
@@ -99,6 +102,13 @@ int main( void )
+#if defined(MBEDTLS_PLATFORM_C)
+    if( ( ret = mbedtls_platform_setup( &platform_ctx ) ) != 0 )
+    {
+        mbedtls_printf( " failed\n  !  mbedtls_platform_setup returned %d\n\n", -ret );
```
We're displaying a negative return value here, yet at line 239 are displaying it without the negative. We should be self-consistent within the same program.

sbutcher-arm commented on programs/ssl/ssl_client2.c:
```diff
@@ -534,6 +537,14 @@ int main( int argc, char *argv[] )
+#if defined(MBEDTLS_PLATFORM_C)
+    if( ( ret = mbedtls_platform_setup( &platform_ctx ) ) != 0 )
+    {
+        mbedtls_printf( " failed\n  !  mbedtls_platform_setup returned %d\n\n", -ret );
```
We're displaying a decimal return value here, yet elsewhere, are displaying it as a hexadecimal. We should be self-consistent within the same program.

sbutcher-arm commented on programs/ssl/ssl_fork_server.c:
```diff
@@ -100,13 +100,23 @@ int main( void )
+#if defined(MBEDTLS_PLATFORM_C)
+    if( ( ret = mbedtls_platform_setup( &platform_ctx ) ) != 0 )
+    {
+        mbedtls_printf( " failed\n  !  mbedtls_platform_setup returned %d\n\n", -ret );
```
We're displaying a negative return value here, yet elsewhere, are displaying it without the negative. We should be self-consistent within the same program.

sbutcher-arm commented on programs/ssl/ssl_mail_client.c:
```diff
@@ -367,6 +370,14 @@ int main( int argc, char *argv[] )
+#if defined(MBEDTLS_PLATFORM_C)
+    if( ( ret = mbedtls_platform_setup( &platform_ctx ) ) != 0 )
+    {
+        mbedtls_printf( " failed\n  !  mbedtls_platform_setup returned %d\n\n", -ret );
```
We're displaying a negative return value here, yet elsewhere, are displaying it without the negative. We should be self-consistent within the same program.

sbutcher-arm commented on programs/ssl/ssl_pthread_server.c:
```diff
@@ -331,6 +334,14 @@ int main( void )
+#if defined(MBEDTLS_PLATFORM_C)
+    if( ( ret = mbedtls_platform_setup( &platform_ctx ) ) != 0 )
+    {
+        mbedtls_printf( " failed\n  !  mbedtls_platform_setup returned %d\n\n", -ret );
```
We're displaying a negative return value here, yet elsewhere are displaying it without the negative. We should be self-consistent within the same program.

sbutcher-arm commented on programs/ssl/ssl_server2.c:
```diff
@@ -957,6 +960,14 @@ int main( int argc, char *argv[] )
+#if defined(MBEDTLS_PLATFORM_C)
+    if( mbedtls_platform_setup( &platform_ctx ) != 0 )
+    {
+        mbedtls_printf( "Failed initializing platform.\n" );
```
We should really display the return code here, similar to other test programs, and other places within this example.

sbutcher-arm commented on programs/test/ssl_cert_test.c:
```diff
@@ -81,10 +81,21 @@ const char *client_private_keys[MAX_CLIENT_CERTS] =
+#if defined(MBEDTLS_PLATFORM_C)
+    if( ( ret = mbedtls_platform_setup( &platform_ctx ) ) != 0 )
+    {
+        mbedtls_printf( " failed\n  !  mbedtls_platform_setup returned %d\n\n", -ret );
```
We're displaying a negative return value here, yet at line 227, are displaying it without the negative. We should be self-consistent within the same program.

sbutcher-arm commented on programs/test/udp_proxy.c:
```diff
@@ -139,18 +139,7 @@ static struct options
     unsigned int seed;          /* seed for "random" events                 */
 } opt;
 
-static void exit_usage( const char *name, const char *value )
```
There is no reason to touch `udp_proxy.c`as far as I can see. If the changes are not needed, let's revert them.

sbutcher-arm commented on programs/util/pem2der.c:
```diff
@@ -280,7 +280,6 @@ int main( int argc, char *argv[] )
     mbedtls_printf( "  + Press Enter to exit this program.\n" );
     fflush( stdout ); getchar();
 #endif
-
```
Removing a single line is unnecessary. Let's revert this change.

sbutcher-arm commented on programs/util/strerror.c:
```diff
@@ -59,7 +59,8 @@ int main( int argc, char *argv[] )
     if( argc != 2 )
     {
         mbedtls_printf( USAGE );
-        return( 0 );
```
There's no need to touch this sample. Let's revert this change.

sbutcher-arm commented on programs/x509/cert_app.c:
```diff
@@ -160,6 +163,14 @@ int main( int argc, char *argv[] )
+#if defined(MBEDTLS_PLATFORM_C)
+    if( ( ret = mbedtls_platform_setup( &platform_ctx ) ) != 0 )
+    {
+        mbedtls_printf( " failed\n  !  mbedtls_platform_setup returned %d\n\n", -ret );
```
We're displaying a negative return value here, yet at line 443 and elsewhere, are displaying it without the negative. We should be self-consistent within the same program.

sbutcher-arm commented on programs/x509/cert_req.c:
```diff
@@ -138,11 +138,22 @@ int main( int argc, char *argv[] )
+#if defined(MBEDTLS_PLATFORM_C)
+    if( ( ret = mbedtls_platform_setup( &platform_ctx ) ) != 0 )
+    {
+        mbedtls_printf( " failed\n  !  mbedtls_platform_setup returned %d\n\n", -ret );
```
We're displaying a negative return value here, yet elsewhere, are displaying it without the negative. We should be self-consistent within the same program.

sbutcher-arm commented on programs/x509/cert_write.c:
```diff
@@ -230,6 +233,14 @@ int main( int argc, char *argv[] )
+#if defined(MBEDTLS_PLATFORM_C)
+    if( ( ret = mbedtls_platform_setup( &platform_ctx ) ) != 0 )
+    {
+        mbedtls_printf( " failed\n  !  mbedtls_platform_setup returned %d\n\n", -ret );
```
We're displaying a negative decimal return value here, yet elsewhere, are displaying it as a hexadecimal. We should be self-consistent within the same program.

sbutcher-arm commented on programs/x509/crl_app.c:
```diff
@@ -72,6 +72,14 @@ int main( int argc, char *argv[] )
+    mbedtls_platform_context platform_ctx;
+    if( ( ret = mbedtls_platform_setup( &platform_ctx ) ) != 0 )
+    {
+        mbedtls_printf( " failed\n  !  mbedtls_platform_setup returned %d\n\n", -ret );
```
We're displaying a negative return value here, yet at line 113, are displaying it without the negative. We should be self-consistent within the same program.

sbutcher-arm commented on programs/x509/req_app.c:
```diff
@@ -72,6 +72,14 @@ int main( int argc, char *argv[] )
+    mbedtls_platform_context platform_ctx;
+    if( ( ret = mbedtls_platform_setup( &platform_ctx ) ) != 0 )
+    {
+        mbedtls_printf( " failed\n  !  mbedtls_platform_setup returned %d\n\n", -ret );
```
We're displaying a negative return value here, yet at line 113, are displaying it without the negative. We should be self-consistent within the same program.

AndrzejKurek commented:
@sbutcher-arm I have chosen to be consistent within the changes i'm introducing, but I think that you're right that all errors in examples should be unified. I would opt for filing a bug ticket and prompt a discussion on whether the errors should be negated or not.

AndrzejKurek commented:
Error unification should be, in my humble opinion, performed in a separate PR across all the example programs, and covers a wider variety of problems. Further description can be found in IOTSSL-2246.
I opt for merging this PR as-is (so that the main changes, platform setup and teardown are merged), followed by a PR fixing all error handling.

sbutcher-arm commented:
@k-stachowiak - PR #532 is now ready for merge. Can you please rebase this PR against that? We can then re-review.

sbutcher-arm commented:
@k-stachowiak - Looks like this needs rebasing again. Could you please rebase to remove the conflicts?

sbutcher-arm commented:
If the rebase is complete and no further changes are needed, can you please the change the label to `needs review`.

k-stachowiak commented:
@sbutcher-arm I was waiting for the CI results before reporting the PR as ready for review. Jenkins reports a failure in a timing test, which I suppose is a known "non-issue", but Travis reports an ambiguous "...exited with 1" error, without an obvious error visible, so I'd like to go through the (rather long) log and look for the issue. Just grepping "error" and "failure" didn't lead to an error, so I'll look at it more closely before checking it as no longer needing work.

k-stachowiak commented:
Nevermind, I found the issue. For some reason tabs got into the source code leading to the failure of the `check-files.py` script. Working on it.

k-stachowiak commented:
@sbutcher-arm @RonEld @AndrzejKurek I rebased the PR and got the result to pass the CI. Could you please re-review the PR?

AndrzejKurek commented on ChangeLog:
```diff
@@ -62,6 +62,9 @@ Changes
      or CBC ciphersuites in (D)TLS versions 1.1 or higher. Fixes #1913, #1914.
    * Add support for buffering of out-of-order handshake messages.
+   * Add calls to platform setup and teardown functions to all the example
+     programs for which this is relevant. This change enables the example
```
I would rephrase the second sentence, as now it suggests that, in short, "this change enables the example programs". How about "This change enables the utilization of plaform-specific cryptographic features in example programs."?

k-stachowiak replied:
I agree the sentence is not nicely worded. AFAIK it would technically be correct to write "...enables the example programs to utilize...", but it is rather informal. I will rephrase the sentence.

AndrzejKurek commented on programs/pkey/key_app.c:
```diff
@@ -292,7 +303,7 @@ int main( int argc, char *argv[] )
 
 #if defined(MBEDTLS_ERROR_C)
-    if( exit_code != MBEDTLS_EXIT_SUCCESS )
+    if( exit_code != MBEDTLS_EXIT_SUCCESS && ret != 0 )
```
Why are both of these tested? Shouldn't these two states correlate with each other?

k-stachowiak commented:
@AndrzejKurek I have addressed your comments. Could you please re-review the PR?

AndrzejKurek commented on programs/pkey/key_app_writer.c:
```diff
@@ -192,15 +192,26 @@ static int write_private_key( mbedtls_pk_context *key, const char *output_file )
+    if( ( ret = mbedtls_platform_setup( &platform_ctx ) ) != 0 )
+    {
+        mbedtls_printf( " failed\n  !  mbedtls_platform_setup returned %d\n", -ret );
+        return( 1 );
```
This place is an example of inconsistent return/exit handling. Please compare this to line 424:
`if( exit_code != MBEDTLS_EXIT_SUCCESS && ret != 0 )`
I can't see a good reason for having 3 places to handle returning an error. 

What I suggest is to get rid of the "exit_code" variable in this whole PR, leave the "ret" variable, and, if we're not overwriting it with MBEDTLS_EXIT_SUCCESS or MBEDTLS_EXIT_FAILURE by hand entirely in the scope of a program (thus passing it outside) - add a bound check in the `exit:` stage.

RonEld commented on programs/pkey/gen_key.c:
```diff
@@ -436,6 +447,10 @@ int main( int argc, char *argv[] )
     fflush( stdout ); getchar();
 #endif
 
+#if defined(MBEDTLS_PLATFORM_C)
```
Any thoughts on this?

k-stachowiak commented:
@AndrzejKurek As discussed outside this PR, due to the reasons:

* consistency of the return code handling being outside the PR's scope,
* examples may be subject to significant changes in the near future,
* the inconsistency not being an issue of correctness, but rather style,

could you please reconsider the current state of the PR acceptable as resolving the main issue - introduction of the setup and teardown code in example programs?

AndrzejKurek commented:
@k-stachowiak as long as the issue I've mentioned is written down somewhere - it's fine by me.

k-stachowiak commented:
@RonEld I agree with that remark and I have just noticed that the rebasing has done some serious mess regarding these teardown commands. I will fix, which will effectively address your comment favourably.

k-stachowiak commented:
retest

k-stachowiak commented:
I have noticed that certain parts of code have been duplicated (thanks to Ron's comment), most probably during a rebase. I have fixed it and, along the line, I addressed @sbutcher-arm 's comments regarding inconsistent error codes printing. Even though we agreed not to do too much work outside the core scope of the PR, I still did it as it didn't cost much, since I had to go through all the programs anyway.

@RonEld @AndrzejKurek Could you please re-review the PR?

AndrzejKurek commented on programs/hash/hello.c:
```diff
@@ -48,7 +48,7 @@ int main( void )
 int main( void )
 {
-    int i, ret;
+    int i, ret = MBEDTLS_EXIT_SUCCESS;
```
I'd rather opt for initializing return codes with invalid values, but that's a minor.

sbutcher-arm commented:
@RonEld - Can you re-review this? We need to close the platform work.

RonEld commented on programs/hash/hello.c:
```diff
@@ -48,25 +48,40 @@ int main( void )
 int main( void )
 {
-    int i, ret;
+    int i, ret = MBEDTLS_EXIT_SUCCESS;
```
Please add a variable for `exit_code` and distinguish between the return code from the functions (named `ret`) and the error code returned to the user( named `exit_code`)

RonEld commented on programs/hash/hello.c:
```diff
@@ -48,25 +48,40 @@ int main( void )
     if( ( ret = mbedtls_md5_ret( (unsigned char *) str, 13, digest ) ) != 0 )
-        return( MBEDTLS_EXIT_FAILURE );
+    {
+        ret = MBEDTLS_EXIT_FAILURE;
```
change to `exit_code`

RonEld commented on programs/hash/hello.c:
```diff
@@ -48,25 +48,40 @@ int main( void )
 #endif
-
-    return( MBEDTLS_EXIT_SUCCESS );
+    return( ret );
```
change to `exit_code`

k-stachowiak commented:
@RonEld Thanks for the review. I addressed your remarks. @sbutcher-arm Upon Ron's remarks I addressed some of your earlier comments. Could you please re-review?

RonEld commented on programs/ssl/dtls_client.c:
```diff
@@ -333,15 +344,17 @@ int main( int argc, char *argv[] )
+    mbedtls_printf( "  + Press Enter to exit this program.\n" );
+    fflush( stdout ); getchar();
+#endif
     return( ret );
```
THis is in the oriignal, but should be exit code

RonEld commented on programs/ssl/ssl_pthread_server.c:
```diff
@@ -514,12 +525,13 @@ int main( void )
     fflush( stdout ); getchar();
 #endif
-
     return( ret );
```
This is in the oroiginal, but should be exit_code. or at least change this value, as the shell cannot handle errors larger than 128

RonEld commented on programs/ssl/ssl_server.c:
```diff
@@ -387,12 +398,13 @@ int main( void )
     fflush( stdout ); getchar();
 #endif
-
     return( ret );
```
This is in the original, but should be exit_code

k-stachowiak commented:
@RonEld Thanks for the further comments. I addressed them, could you please re-review?

sbutcher-arm commented on programs/x509/cert_app.c:
```diff
@@ -163,6 +166,14 @@ int main( int argc, char *argv[] )
+    if( ( ret = mbedtls_platform_setup( &platform_ctx ) ) != 0 )
+    {
+        mbedtls_printf( " failed\n  !  mbedtls_platform_setup returned -0x%x\n\n", -ret );
+        return( 1 );
```
Should be:
```
mbedtls_exit( MBEDTLS_EXIT_FAILURE );
```

sbutcher-arm commented on programs/x509/cert_req.c:
```diff
@@ -142,11 +142,22 @@ int main( int argc, char *argv[] )
+    if( ( ret = mbedtls_platform_setup( &platform_ctx ) ) != 0 )
+    {
+        mbedtls_printf( " failed\n  !  mbedtls_platform_setup returned %d\n\n", ret );
+        return( 1 );
```
Should be:
```
mbedtls_exit( MBEDTLS_EXIT_FAILURE );
```

sbutcher-arm commented on programs/x509/cert_write.c:
```diff
@@ -234,6 +237,14 @@ int main( int argc, char *argv[] )
+    if( ( ret = mbedtls_platform_setup( &platform_ctx ) ) != 0 )
+    {
+        mbedtls_printf( " failed\n  !  mbedtls_platform_setup returned -0x%04x\n\n", -ret );
+        return( 1 );
```
Should be:
```
mbedtls_exit( MBEDTLS_EXIT_FAILURE );
```

sbutcher-arm commented on programs/x509/crl_app.c:
```diff
@@ -76,6 +76,14 @@ int main( int argc, char *argv[] )
+    if( ( ret = mbedtls_platform_setup( &platform_ctx ) ) != 0 )
+    {
+        mbedtls_printf( " failed\n  !  mbedtls_platform_setup returned %d\n\n", ret );
+        return( 1 );
```
Should be:
```
mbedtls_exit( MBEDTLS_EXIT_FAILURE );
```

sbutcher-arm commented on programs/x509/req_app.c:
```diff
@@ -76,6 +76,14 @@ int main( int argc, char *argv[] )
+    if( ( ret = mbedtls_platform_setup( &platform_ctx ) ) != 0 )
+    {
+        mbedtls_printf( " failed\n  !  mbedtls_platform_setup returned %d\n\n", ret );
+        return( 1 );
```
Should be:
```
mbedtls_exit( MBEDTLS_EXIT_FAILURE );
```

k-stachowiak commented:
retest

RonEld commented on ChangeLog:
```diff
@@ -62,9 +62,9 @@ Changes
-     programs for which this is relevant. This change enables the utilization
-     of platform-specific cryptographic features in example programs.
+   * Add to the example programs calls to platform initialisation and teardown
+     functions, which can be optionally used by platform specific code for
```
I think it is better to rephrase:
"... which can be optionally implemented by platform ..."

k-stachowiak commented:
@RonEld Thanks for the review! I addressed your comment.

k-stachowiak commented on programs/x509/cert_write.c:
```diff
@@ -795,7 +806,9 @@ int main( int argc, char *argv[] )
     mbedtls_entropy_free( &entropy );
-
+#if defined(MBEDTLS_PLATFORM_C)
+    mbedtls_platform_teardown( &platform_ctx );
```
I made some changes regarding this remark at the beginning of several programs. However, I would like to leave the windows-specific endings after the platform teardown. This has been done this way upon @RonEld's request expressed here: https://github.com/ARMmbed/mbedtls/pull/1572#discussion_r181968937. I think it is reasonable on Windows (and currently limited to that platform), and I would not expect `printf` to have platform specific behavior there.

k-stachowiak commented:
@RonEld Thanks for a thorough review. I would gladly fix the commit message if it didn't require history rewriting in some way.

sbutcher-arm commented:
@k-stachowiak - I hate to ask - but can you please rebase?

k-stachowiak commented:
@sbutcher-arm @RonEld I have rebased the PR, could you please re-review it? As there was a significant interference from other changes, I pushed the original branch here: https://github.com/k-stachowiak/mbedtls/tree/platform-setup-terminate_2019-01-08 for comparison.

sbutcher-arm commented:
Travis is failing:
```
[ 34%] [32mBuilding C object programs/ssl/CMakeFiles/ssl_client1.dir/ssl_client1.c.o[0m
[ 35%] [32m[1mLinking C executable ssl_client1[0m
[ 35%] Built target ssl_client1
[35m[1mScanning dependencies of target dtls_server[0m
[ 35%] [32mBuilding C object programs/ssl/CMakeFiles/dtls_server.dir/dtls_server.c.o[0m
[1m/home/travis/build/ARMmbed/mbedtls/programs/ssl/dtls_server.c:64:5: [0m[0;1;31merror: [0m[1m
      implicitly declaring library function 'exit' with type 'void (int)
      __attribute__((noreturn))' [-Werror,-Wimplicit-function-declaration][0m
    mbedtls_exit( 0 );
[0;1;32m    ^
[0m[1m/home/travis/build/ARMmbed/mbedtls/programs/ssl/dtls_server.c:35:33: [0m[0;1;30mnote: [0m
      expanded from macro 'mbedtls_exit'[0m
#define mbedtls_exit            exit
[0;1;32m                                ^
[0m[1m/home/travis/build/ARMmbed/mbedtls/programs/ssl/dtls_server.c:64:5: [0m[0;1;30mnote: [0m
      include the header <stdlib.h> or explicitly provide a declaration for
      'exit'[0m
[1m/home/travis/build/ARMmbed/mbedtls/programs/ssl/dtls_server.c:35:33: [0m[0;1;30mnote: [0m
      expanded from macro 'mbedtls_exit'[0m
#define mbedtls_exit            exit
[0;1;32m                                ^
[0m1 error generated.
make[2]: *** [programs/ssl/CMakeFiles/dtls_server.dir/dtls_server.c.o] Error 1
make[1]: *** [programs/ssl/CMakeFiles/dtls_server.dir/all] Error 2
make: *** [all] Error 2
Failed to build: config-mini-tls1_1.h
travis_time:end:091a3394:start=1546955191796030193,finish=1546955201068264944,duration=9272234751
[0K[31;1mThe command "tests/scripts/test-ref-configs.pl" exited with 1.[0m
```

k-stachowiak commented:
@sbutcher-arm Only the freebsd timing tests are failing now.

sbutcher-arm commented:
@k-stachowiak - I hate to ask (yet again) but can you rebase, so we can get this PR in?

k-stachowiak commented:
@sbutcher-arm I have rebased this PR

sbutcher-arm commented on programs/pkey/rsa_decrypt.c:
```diff
@@ -89,15 +93,22 @@ int main( int argc, char *argv[] )
+    if( ( ret = mbedtls_platform_setup( &platform_ctx ) ) != 0 )
+    {
+        mbedtls_fprintf(
+            stderr, "platofrm_setup returned %d\n", ret );
```
'platofrm_setup' should be 'platform_setup'

k-stachowiak commented:
@RonEld Thanks for that out and an additional off-line discussion. I went through the programs and found one instance, where I am certain that I made a mistake. I have pushed a fix for that.
Otherwise, the output stream choices, even if arguable, were deliberate.

RonEld commented:
Have you run `ssl-opts.sh`? AFAIK, ssl-opts filters stdout for specific messages. Isn't this problematic?

k-stachowiak commented:
@sbutcher-arm @RonEld I have reverted some changes made to the output of the programs used in `ssl-opt.sh` and `compat.sh`, so that they produce expected strings in the expected streams. Can you please re-review the PR? If the changes are approved, I will backport them. Thanks in advance!

k-stachowiak commented:
The most recent forced push was an amendment of only the most recent commit.

k-stachowiak commented:
Unfortunately the CI is failing with some, seemingly unrelated errors. There is at least:

```
Last error was: -0xFFFFFFFE - BIGNUM - An error occurred while reading from or writing to a file
```

being reported.

k-stachowiak commented:
I'm getting a sense of hopelessness from this PR, therefore I will start raising smaller ones to take some changes off this one. Ultimately I'd like to:

1) Gather all the changes in this PR that exceed its scope
2) Introduce the out of scope changes in other PRs
3) Rebase and reset this PR after the out of scope changes have got into development
4) At this point, this PR should only contain the platform initialization related changes and should be far easier to review and, finally, merge.

The first side-PR is: #2595

tom-cosgrove-arm commented:
Closing this PR - as per Krzysztof's comment above it is unwieldy, there are now many conflicts, and much of the work has now been done in smaller PRs
